source/postprocess/geoid.cc

```diff
@@ -0,0 +1,975 @@
+        AssertThrow( a.size() == max_degree+1 , ExcInternalError() );
+
+        for (unsigned int l = 2, k = 0; l <= max_degree; ++l)
+          for (unsigned int m = 0; m <= l; ++m, ++k)
+            {
+              coefficients.sine_coefficients[k] = s[l] * coefficients.sine_coefficients[k] +
+                                                  a[l] * M.get_coefficients().sine_coefficients[k];
+              coefficients.cosine_coefficients[k] = s[l] * coefficients.cosine_coefficients[k] +
+                                                    a[l] * M.get_coefficients().cosine_coefficients[k];
+            }
+      }
+
+      template <int dim>
+      void
+      MultipoleExpansion<dim>::mpi_sum_coefficients (MPI_Comm mpi_communicator)
+      {
+        dealii::Utilities::MPI::sum(coefficients.sine_coefficients,mpi_communicator,coefficients.sine_coefficients);
+        dealii::Utilities::MPI::sum(coefficients.cosine_coefficients,mpi_communicator,coefficients.cosine_coefficients);
+      }
+
+
+
+      /* Implementations for doing an external multipole expansion */
+
+      template <int dim>
+      ExternalMultipoleExpansion<dim>::ExternalMultipoleExpansion(const unsigned int max_degree,
+                                                                  const double evaluation_radius)
+        : MultipoleExpansion<dim>(max_degree),
+          evaluation_radius(evaluation_radius)
+      {}
+
+      template <>
+      void
+      ExternalMultipoleExpansion<2>::add_quadrature_point (const Point<2> &position,
+                                                           const double value,
+                                                           const double weight)
+      {
+        const double r = position.norm();
+        const double theta = std::atan2(position[1],position[0]);
+
+        Assert(r/evaluation_radius -1. <= 1.e-7 ,
+               ExcMessage("Can't do an external multipole expansion with this radius") );
+
+        if ( evaluation_radius > 0.)
+          {
+            for (unsigned int n = 2; n <= max_degree; ++n)
+              {
+                const double factor = value * std::pow( r / evaluation_radius, static_cast<double>(n) )
+                                      / static_cast<double>(n) * weight;
+                coefficients.cosine_coefficients[n] += factor * std::cos( static_cast<double>(n) * theta);
+                coefficients.sine_coefficients[n] +=   factor * std::sin( static_cast<double>(n) * theta);
+              }
+          }
+      }
+
+      template <>
+      void
+      ExternalMultipoleExpansion<3>::add_quadrature_point (const Point<3> &position,
+                                                           const double value,
+                                                           const double weight)
+      {
+        const double r = position.norm();
+        const double phi = std::atan2(position[1],position[0]);
+        const double theta = std::acos(position[2]/r);
```

Comment:
Please use the Utilities::spherical_coordinates function for the computation to prevent code duplication.